### PR TITLE
Allow different jdbc metadata cache ttls for schema and tables lists

### PR DESCRIPTION
**Pull request by atanasenko**
## Description
Missing table metadata can be configured to not get cached (`cacheMissing`), but that doesn't apply to schema and table name lists.

In certain cases it might be beneficial to shorten the list cache ttl for them to reflect the added schemas/tables faster, while specific table caches will automatically reload tables that were missing before.

## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

**Comment by hashhar** [comment]
This is making configuration very complex over time. What is the actual problem we are trying to solve? Do we want to notice newly created objects faster? Or is the goal something else?

**Comment by hashhar** [comment]
> Missing table metadata can be configured to not get cached (cacheMissing), but that doesn't apply to schema and table name lists.

To me it sounds like this is the solution we should implement - it would solve the problem in a simpler way and also make the config apply consistently.

**Comment by skrzypo987** [comment]
> To me it sounds like this is the solution we should implement

How should that work? The list of objects is not a 0/1 situation. With a table it either exists or it doesn't. A list is a list. It doesn't show up when the table is created, but it is changed.

**Review by skrzypo987 — COMMENTED** [review]
Still some tests failing.
Other than that looks good

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcConfig.java [hunk 1]
```diff
@@ -117,9 +150,19 @@ public BaseJdbcConfig setCacheMaximumSize(long cacheMaximumSize)
...
+                        format("%s must be set to a non-zero value when %s is set", METADATA_CACHE_TTL, METADATA_CACHE_MAXIMUM_SIZE));
+            }
+            if (schemaNamesCacheTtl.isPresent() && !schemaNamesCacheTtl.get().equals(CACHING_DISABLED)) {
+                throw new IllegalArgumentException(
+                        format("%s must not be set when %s is not set", METADATA_SCHEMAS_CACHE_TTL, METADATA_CACHE_TTL));
+            }
+            if (tableNamesCacheTtl.isPresent() && !tableNamesCacheTtl.get().equals(CACHING_DISABLED)) {
+                throw new IllegalArgumentException(
+                        format("%s must not be set when %s is not set", METADATA_TABLES_CACHE_TTL, METADATA_CACHE_TTL));
+            }
```
Comment (Praveen2112) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcConfig.java:165]: nit: Can we expose them as additional methods with `AssertTrue` annotation ? One for schemaName cache and one for tableName cache ?

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/CachingJdbcClient.java [hunk 2]
```diff
@@ -111,25 +140,26 @@ public CachingJdbcClient(
...
-        }
-
-        schemaNamesCache = cacheBuilder.build();
-        tableNamesCache = cacheBuilder.build();
-        tableHandlesByNameCache = cacheBuilder.build();
-        tableHandlesByQueryCache = cacheBuilder.build();
-        columnsCache = cacheBuilder.build();
-        statisticsCache = cacheBuilder.build();
+        schemaNamesCache = cacheBuilder
+                .apply(schemaNamesCachingTtl.orElse(metadataCachingTtl))
```
Comment (Praveen2112) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/CachingJdbcClient.java:154]: nit: Can we move this logic to config object ? instead of having it here ?

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/CachingJdbcClient.java [hunk 3]
```diff
@@ -583,6 +613,12 @@ private void invalidateColumnsCache(SchemaTableName table)
         invalidateCache(columnsCache, key -> key.table.equals(table));
     }
 
+    @VisibleForTesting
```
Comment (Praveen2112) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/CachingJdbcClient.java:616]: Can we have this change as a separate commit ? As it allows Trino to expose schema stats ?

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/CachingJdbcClient.java [hunk 4]
```diff
@@ -103,6 +112,26 @@ public CachingJdbcClient(
...
+    {
+        this(delegate,
+                sessionPropertiesProviders,
+                identityMapping,
+                metadataCachingTtl,
+                Optional.empty(),
+                Optional.empty(),
+                cacheMissing,
+                cacheMaximumSize);
+    }
```
Comment (Praveen2112) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/CachingJdbcClient.java:124]: Can we remove this constructor use the other one at all places ?

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/CachingJdbcClient.java (hunk 3 above)
Comment (Praveen2112) [reply plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/CachingJdbcClient.java:616]: And add some test for schema stats ?

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java [hunk 5]
```diff
@@ -294,10 +307,10 @@ private void assertTableHandlesByNameCacheIsInvalidated(JdbcTableHandle table)
     {
         SchemaTableName tableName = table.asPlainTable().getSchemaTableName();
 
-        assertCacheStats(cachingJdbcClient, TABLE_HANDLES_BY_NAME_CACHE).misses(1).loads(1).afterRunning(() -> {
+        assertTableHandleCache(cachingJdbcClient).misses(1).loads(1).afterRunning(() -> {
             assertThat(cachingJdbcClient.getTableHandle(SESSION, tableName).orElseThrow()).isEqualTo(table);
         });
-        assertCacheStats(cachingJdbcClient, TABLE_HANDLES_BY_NAME_CACHE).hits(1).afterRunning(() -> {
+        assertTableHandleCache(cachingJdbcClient).hits(1).afterRunning(() -> {
```
Comment (Praveen2112) [inline plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java:313]: Can this cleanup be a different commit as a cleanup ?

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java [hunk 6]
```diff
@@ -838,6 +863,119 @@ public Optional<JdbcTableHandle> getTableHandle(ConnectorSession session, Schema
                         "com.google.common.util.concurrent.UncheckedExecutionException: java.lang.RuntimeException: first attempt is poised to fail");
     }
 
+    @Test
+    public void testSpecificSchemaAndTableCachesWithCacheMissing()
+            throws Exception
```
Comment (Praveen2112) [inline plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java:868]: we can remove the `throws`

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java [hunk 7]
```diff
@@ -838,6 +863,119 @@ public Optional<JdbcTableHandle> getTableHandle(ConnectorSession session, Schema
...
+    {
+        assertSpecificSchemaAndTableCaches(true);
+    }
+
+    @Test
+    public void testSpecificSchemaAndTableCachesWithoutCacheMissing()
+            throws Exception
```
Comment (Praveen2112) [inline plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java:875]: ditto

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java [hunk 8]
```diff
@@ -838,6 +863,119 @@ public Optional<JdbcTableHandle> getTableHandle(ConnectorSession session, Schema
...
+    {
+        assertSpecificSchemaAndTableCaches(false);
+    }
+
+    private void assertSpecificSchemaAndTableCaches(boolean cacheMissing)
```
Comment (Praveen2112) [inline plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java:880]: Can we split this into two methods - one for table and other for schema ?

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java [hunk 9]
```diff
@@ -838,6 +863,119 @@ public Optional<JdbcTableHandle> getTableHandle(ConnectorSession session, Schema
...
+                    .doesNotContain(secondName);
+        });
+        assertTableHandleCache(cachingJdbcClient).hits(1).afterRunning(() -> {
+            assertThat(cachingJdbcClient.getTableHandle(session, firstName)).isNotEmpty();
+        });
+        if (cacheMissing) {
+            assertTableHandleCache(cachingJdbcClient).hits(1).afterRunning(() -> {
+                assertThat(cachingJdbcClient.getTableHandle(session, secondName)).isEmpty();
+            });
+        }
```
Comment (Praveen2112) [inline plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java:922]: Can we have dedicated test with cache disabled ?

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/CachingJdbcClient.java (hunk 4 above)
Comment (atanasenko) [reply plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/CachingJdbcClient.java]: Those options are now non-optional and will have to be equal to `metadataCachintTtl` in such cases. I think it's easier to have an overloaded constructor to indicate that.

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java (hunk 8 above)
Comment (atanasenko) [reply plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java:896]: The test ensures that schema and table names cache ttls are different, it won't do that if they are separate.

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java (hunk 9 above)
Comment (atanasenko) [reply plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java]: The difference is pretty small, but the method is large. It will be hard to spot differences if there will be two different flavors of it.

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java [hunk 10]
```diff
@@ -891,19 +906,34 @@ public void testEverythingImplemented()
...
+
     private static SingleJdbcCacheStatsAssertions assertColumnCacheStats(CachingJdbcClient client)
     {
         return assertCacheStats(client, CachingJdbcCache.COLUMNS_CACHE);
     }
 
     private static SingleJdbcCacheStatsAssertions assertStatisticsCacheStats(CachingJdbcClient client)
     {
-        return assertCacheStats(client, STATISTICS_CACHE);
+        return assertCacheStats(client, CachingJdbcCache.STATISTICS_CACHE);
```
Comment (Praveen2112) [inline plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java:1075]: This one is kind of an odd-one here in this commit.

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/CachingJdbcClient.java [hunk 11]
```diff
@@ -583,6 +583,12 @@ private void invalidateColumnsCache(SchemaTableName table)
         invalidateCache(columnsCache, key -> key.table.equals(table));
     }
 
+    @VisibleForTesting
+    CacheStats getSchemaNamesCacheStats()
```
Comment (Praveen2112) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/CachingJdbcClient.java:613]: We can split this commit into two - First we could have a commit which `Expose SchemName cache stats` which expose the stats and a corresponding testing. Then a commit which extracts table handler assertions. WDYT ?

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcConfig.java [hunk 12]
```diff
@@ -114,12 +146,21 @@ public BaseJdbcConfig setCacheMaximumSize(long cacheMaximumSize)
         return this;
     }
 
-    @PostConstruct
-    public void validate()
```
Comment (Praveen2112) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcConfig.java:118]: Changes related to `validate` can be in a different commit or we could change them in a follow-up.

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcConfig.java [hunk 13]
```diff
@@ -114,12 +146,21 @@ public BaseJdbcConfig setCacheMaximumSize(long cacheMaximumSize)
...
+    @AssertTrue(message = METADATA_CACHE_TTL + " must be set to a non-zero value when " + METADATA_CACHE_MAXIMUM_SIZE + " is set")
+    public boolean isCacheMaximumSizeConsistent()
+    {
+        return metadataCacheTtl.equals(CACHING_DISABLED) ? cacheMaximumSize == BaseJdbcConfig.DEFAULT_METADATA_CACHE_SIZE : true;
+    }
+
+    @AssertTrue(message = METADATA_SCHEMAS_CACHE_TTL + " must not be set when " + METADATA_CACHE_TTL + " is not set")
+    public boolean isSchemaNamesCacheTtlConsistent()
```
Comment (Praveen2112) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcConfig.java:156]: `schemaNamesCacheTtlValidation`. WDYT ?

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcConfig.java [hunk 14]
```diff
@@ -114,12 +146,21 @@ public BaseJdbcConfig setCacheMaximumSize(long cacheMaximumSize)
...
+    {
+        return metadataCacheTtl.equals(CACHING_DISABLED) ? schemaNamesCacheTtl.isEmpty() : true;
```
Comment (Praveen2112) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcConfig.java:158]: Can we replace with `AssertFalse` and have a condition like `metadataCacheTtl.equals(CACHING_DISABLED) && schemaNamesCacheTtl.isPresent()` or have `AssertTrue` with `!metadataCacheTtl.equals(CACHING_DISABLED)  || schemaNamesCacheTtl.isEmpty()`

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/CachingJdbcClient.java [hunk 15]
```diff
@@ -111,25 +140,22 @@ public CachingJdbcClient(
         this.cacheMissing = cacheMissing;
         this.identityMapping = requireNonNull(identityMapping, "identityMapping is null");
 
-        EvictableCacheBuilder<Object, Object> cacheBuilder = EvictableCacheBuilder.newBuilder()
-                .expireAfterWrite(metadataCachingTtl.toMillis(), MILLISECONDS)
+        long cacheSize = metadataCachingTtl.equals(CACHING_DISABLED)
+                // Disables the cache entirely
+                ? 0
+                : cacheMaximumSize;
+        Function<Duration, EvictableCacheBuilder<Object, Object>> cacheBuilder = cachingTtl -> EvictableCacheBuilder.newBuilder()
```
Comment (Praveen2112) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/CachingJdbcClient.java:147]: Instead can we have a `Supplier` or just have builder and share them, since we are initializing a builder - maybe we could build schemaNamesCache and tableNamesCache separately once all the other caches are build - with custom ttl. WDYT ?

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java [hunk 16]
```diff
@@ -654,7 +676,13 @@ public TableStatistics getTableStatistics(ConnectorSession session, JdbcTableHan
                 return NON_EMPTY_STATS;
             }
         };
-        return new CachingJdbcClient(statsAwareJdbcClient, SESSION_PROPERTIES_PROVIDERS, new SingletonIdentityCacheMapping(), duration, cacheMissing, cacheMaximumSize);
+        return new CachingJdbcClient(
```
Comment (Praveen2112) [inline plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java:679]: Can we revert this or have a dedicated commit for these re-arrangement.

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java (hunk 8 above)
Comment (Praveen2112) [reply plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java:896]: The test method is quite big and it does assert multiple operations like with and without cache. Or we could have schema name cache, table name cache and a test method with combined configuration.

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java (hunk 10 above)
Comment (atanasenko) [reply plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java:1075]: Following the same approach as other methods.

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcConfig.java (hunk 13 above)
Comment (atanasenko) [reply plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcConfig.java:156]: `ValidationAssertions#assertFailsValidation()` tests those methods by property name it represents.

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java (hunk 8 above)
Comment (atanasenko) [reply plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java:896]: Those calls are inter-dependendent. First of all the test ensure that schema and table caches are different, an then it tests the evolution of the cache over time. I can split it into separate 'stages' which should make it more readable.

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/CachingJdbcClient.java (hunk 15 above)
Comment (atanasenko) [reply plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/CachingJdbcClient.java:147]: Builder cannot be reused, those methods have to be called no more than once.

Comment (atanasenko) [reply plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/CachingJdbcClient.java]: Extracted to a separate method

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java (hunk 8 above)
Comment (atanasenko) [reply plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java:890]: Split it into multiple smaller sections and extracted to a separate assertions class

**Review by skrzypo987 — APPROVED** [review]

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java [hunk 17]
```diff
@@ -1071,4 +1131,111 @@ public <T> T calling(Callable<T> callable)
...
+            });
+            assertTableNamesCache(cachingJdbcClient).hits(1).afterRunning(() -> {
+                assertThat(cachingJdbcClient.getTableNames(session, Optional.empty()))
+                        .contains(firstName)
+                        .doesNotContain(secondName);
+            });
+            assertTableHandleByNameCache(cachingJdbcClient).hits(1).afterRunning(() -> {
+                assertThat(cachingJdbcClient.getTableHandle(session, firstName)).isNotEmpty();
+            });
+            if (cacheMissing) {
```
Comment (Praveen2112) [inline plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java:1179]: Since `cacheMissing` works only at a table level - do we need to mix up with the schema related tests. We could configure some default value of schema cache ttl and create a JDBC client. WDYT ?

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcConfig.java [hunk 18]
```diff
@@ -114,12 +114,10 @@ public BaseJdbcConfig setCacheMaximumSize(long cacheMaximumSize)
...
-    @PostConstruct
-    public void validate()
+    @AssertTrue(message = METADATA_CACHE_TTL + " must be set to a non-zero value when " + METADATA_CACHE_MAXIMUM_SIZE + " is set")
+    public boolean isCacheMaximumSizeConsistent()
     {
-        if (metadataCacheTtl.equals(CACHING_DISABLED) && cacheMaximumSize != BaseJdbcConfig.DEFAULT_METADATA_CACHE_SIZE) {
-            throw new IllegalArgumentException(
-                    format("%s must be set to a non-zero value when %s is set", METADATA_CACHE_TTL, METADATA_CACHE_MAXIMUM_SIZE));
-        }
+        return !metadataCacheTtl.equals(CACHING_DISABLED) || cacheMaximumSize == BaseJdbcConfig.DEFAULT_METADATA_CACHE_SIZE;
```
[the remark this answers is not in the document]
Comment (atanasenko) [reply plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcConfig.java:152]: Not sure what you mean here.

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/CachingJdbcClient.java (hunk 15 above)
Comment (atanasenko) [reply plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/CachingJdbcClient.java]: Supplier will mean we need to call the long `expireAfterWrite()` and repeat all the unit conversions multiple times.

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcConfig.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (Praveen2112) [reply plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcConfig.java:152]: I meant either we can have 
```
@AssertFalse
public boolean ...
{
    return metadataCacheTtl.equals(CACHING_DISABLED) && cacheMaximumSize != BaseJdbcConfig.DEFAULT_METADATA_CACHE_SIZE;
}
```

Or we could have 
```
@AssertTrue
public boolean ...
{
    return cacheMaximumSize == BaseJdbcConfig.DEFAULT_METADATA_CACHE_SIZE || !metadataCacheTtl.equals(CACHING_DISABLED);
}
```
Similarly for the next commit

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java (hunk 8 above)
Comment (atanasenko) [reply plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java:890]: Not sure if the two values for boolean warrant a DataProvider, and it's not really data, it's two modes of operations and they should be expressed explcitly.

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcConfig.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (atanasenko) [reply plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcConfig.java:152]: I meant the part about `having the same condition as if`. Also, what would be the difference in swapping the conditions?
Currently those validations are basically saying 'if cache is disabled, check that the <value> is not set`.

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java (hunk 17 above)
Comment (atanasenko) [reply plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java:1179]: I'm testing a usecase where a newly added table gets reflected in metadata lists faster than the general table cache ttl.

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java [hunk 19]
```diff
@@ -1071,4 +1127,97 @@ public <T> T calling(Callable<T> callable)
...
+            assertTableNamesCache(cachingJdbcClient).loads(1).misses(1).afterRunning(() -> {
+                assertThat(cachingJdbcClient.getTableNames(session, Optional.empty()))
+                        .contains(firstName)
+                        .doesNotContain(secondName);
+            });
+            assertTableHandleByNameCache(cachingJdbcClient).misses(1).loads(1).afterRunning(() -> {
+                assertThat(cachingJdbcClient.getTableHandle(session, firstName)).isNotEmpty();
+            });
+            assertTableHandleByNameCache(cachingJdbcClient).misses(1).loads(1).afterRunning(() -> {
+                assertThat(cachingJdbcClient.getTableHandle(session, secondName)).isEmpty();
```
Comment (Praveen2112) [inline plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java:1162]: If we are asserting only the schema and table names do we need this assertion related to `tableHandleByName`.

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java [hunk 20]
```diff
@@ -853,6 +875,40 @@ public Optional<JdbcTableHandle> getTableHandle(ConnectorSession session, Schema
                         "com.google.common.util.concurrent.UncheckedExecutionException: java.lang.RuntimeException: first attempt is poised to fail");
     }
 
+    @Test
+    public void testSpecificSchemaAndTableCaches()
+    {
+        CachingJdbcClient cachingJdbcClient = createCachingJdbcClient(
+                FOREVER,
+                Duration.succinctDuration(2, SECONDS),
+                Duration.succinctDuration(1, SECONDS),
```
Comment (Praveen2112) [inline plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java:884]: Can we increase the difference between the two ttl a bit more higher so as to avoid flakkiness ?

**Comment by findepi** [comment]
@atanasenko please fork with @sopel39 to ensure this is consistent with recent Hive connector metadata caching changes design

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java (hunk 19 above)
Comment (atanasenko) [reply plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java]: It's needed to make sure that metadata caches have separate ttl from individual table caches.

**Comment by atanasenko** [comment]
> Hive connector metadata caching changes design
@findepi those changes seem to be related to extracting a separate ttl for stats, but here I'm separating some of the metadata, cache and individual table metadata are untouched.

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcConfig.java [hunk 21]
```diff
@@ -114,12 +113,10 @@ public BaseJdbcConfig setCacheMaximumSize(long cacheMaximumSize)
...
     }
+
```
Comment (Praveen2112) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcConfig.java:154]: This line can be removed.

File: plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java [hunk 22]
```diff
@@ -853,6 +875,103 @@ public Optional<JdbcTableHandle> getTableHandle(ConnectorSession session, Schema
...
+                assertThat(cachingJdbcClient.getTableHandle(session, firstName)).isNotEmpty();
+            });
+            assertTableHandleByNameCache(cachingJdbcClient).hits(1).afterRunning(() -> {
+                assertThat(cachingJdbcClient.getTableHandle(session, secondName)).isNotEmpty();
+            });
+        });
+
+        jdbcClient.dropTable(SESSION, first);
+        jdbcClient.dropTable(SESSION, second);
+        jdbcClient.dropSchema(SESSION, secondName.getSchemaName());
```
Comment (Praveen2112) [inline plugin/trino-base-jdbc/src/test/java/io/trino/plugin/jdbc/TestCachingJdbcClient.java:972]: nit: `jdbcClient.dropSchema(SESSION, secondSchema)`

**Review by Praveen2112 — APPROVED** [review]
LGTM.

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcConfig.java [hunk 23]
```diff
@@ -86,6 +91,34 @@ public BaseJdbcConfig setMetadataCacheTtl(Duration metadataCacheTtl)
         return this;
     }
 
+    @NotNull
+    public Duration getSchemaNamesCacheTtl()
```
Comment (hashhar) [inline plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcConfig.java:95]: Does it make sense to cap this to be lower than metadata cache ttl? Listing operations will see a table but there would be no metadata cached for it and the table might've been dropped already.

**Review by hashhar — COMMENTED** [review]
skimmed, will do another pass

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcConfig.java (hunk 23 above)
Comment (atanasenko) [reply plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcConfig.java:95]: Lists and specific metadata objects are loaded at different times too, sot this scenario can still happen.

**Review by ssheikin — APPROVED** [review]

**Review by hashhar — APPROVED** [review]
LGTM % whether we want to cap the list cache TTL to be lower than metadata cache ttl

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcConfig.java (hunk 23 above)
Comment (atanasenko) [reply plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcConfig.java:95]: @hashhar Theoretically, there could be some situations when users want to cache lists of schemas/tables for a longer amount of time (very big number of entries which do not change over time), but still let individual table data get reloaded.
This is a generic config after all.

Comment (hashhar) [reply plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcConfig.java:95]: That's a fair point.

**Review by hashhar — APPROVED** [review]
LGTM from me.

**Review by kokosing — APPROVED** [review]

**Comment by colebow** [comment]
Does this need release notes? cc @kokosing

**Comment by kokosing** [comment]
Yes.

```
JDBC connectors
*  Introduce `metadata.schemas.cache-ttl` and `metadata.tables.cache-ttl` configuration properties to control how long metadata information about tables and schemas should be kept in cache memory
```
